### PR TITLE
Add CoffFile::LoadSymbolsFromExportTable

### PR DESCRIPTION
**Pull request by dpallotti**
Similarly to `ElfFile::LoadSymbolsFromDynsym`.
As the name suggests, this extracts symbols from the Export Table. As the Export Table doesn't contain sizes (just like the COFF symbol table), these are obtained from unwind information when possible.
It also supports symbols exported by ordinal only (with no name), to which we arbitrarily assign the name `NONAME<ordinal>`.

Bug: http://b/245680119

Test:
- Unit tests.
- As part of a larger prototype for symbol loading fallback.

### CODE REVIEW

**Comment by dpallotti** [comment]
Hmm, the test `LoadSymbolsFromExportTableAllExportedOnlyByOrdinal` aborts on clang11 debug builds, I can reproduce locally.

**Comment by dpallotti** [comment]
> Hmm, the test `LoadSymbolsFromExportTableAllExportedOnlyByOrdinal` aborts on clang11 debug builds, I can reproduce locally.

Ok, I hadn't fully understood how strict `llvm::Error` is. Might be fixed now.

File: src/ObjectUtils/CoffFile.cpp [hunk 1]
```diff
@@ -366,6 +369,94 @@ bool CoffFileImpl::HasDebugSymbols() const {
...
+  }
+
+  ModuleSymbols module_symbols;
+  for (const llvm::object::ExportDirectoryEntryRef& ref : object_file_->export_directories()) {
+    // From https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#export-address-table:
+    // "A forwarder RVA exports a definition from some other image, making it appear as if it were
+    // being exported by the current image." So these symbols don't really belong to this file, and
+    // we skip them.
+    bool is_forwarder{};
+    if (llvm::Error error = ref.isForwarder(is_forwarder)) {
```
Comment (florian-kuebler) [inline src/ObjectUtils/CoffFile.cpp:396]: It actually confused me quite a bit, that `isForwarder` was taking a reference to write the output. Maybe add a super brief comment?

File: src/ObjectUtils/CoffFile.cpp [hunk 2]
```diff
@@ -366,6 +369,94 @@ bool CoffFileImpl::HasDebugSymbols() const {
...
+
+    SymbolInfo symbol_info;
+    symbol_info.set_demangled_name(std::move(name));
+    symbol_info.set_address(virtual_address);
+
+    // The Export Table doesn't contain the size of symbols: obtain this size from the Exception
+    // Table, but note that that's not always possible, as leaf functions have no RUNTIME_FUNCTION.
+    auto unwind_range_start_to_size_it = unwind_range_start_to_size.find(virtual_address);
+    if (unwind_range_start_to_size_it == unwind_range_start_to_size.end()) {
+      symbol_info.set_size(0);
```
Comment (florian-kuebler) [inline src/ObjectUtils/CoffFile.cpp:439]: I am wondering if we should actually do that. If we don't have a size, we won't be able to do much with the symbol. So we might want to discard it. WDYT?

**Review by florian-kuebler — COMMENTED** [review]
My only question would be about symbols with size 0.
Besides that, LGTM.

File: src/ObjectUtils/CoffFile.cpp (hunk 1 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp:397]: Ok, I added it here. But note that this is the interface of the LLVM method and we can't really add such a comment everywhere this happens.

File: src/ObjectUtils/CoffFile.cpp (hunk 2 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp:440]: First of all, this is consistent with `DeduceDebugSymbolMissingSizesAsDistanceFromNextSymbol`, where we set the size to 0 if it's still unknown: this way at least the symbol shows up at all in the function list, even if its size is wrong.

However, the practical reason is that the idea is that these will be merged with the symbols from unwind info. Once that's the case, we can set these zeros back to `kUnknownSymbolSize`, call `DeduceDebugSymbolMissingSizesAsDistanceFromNextSymbol`, and get mostly correct sizes (unless there's a big gap between an exported symbol and the following symbol from the unwind info, or there's a leaf function without unwind info after the exported symbol).

Comment (florian-kuebler) [reply src/ObjectUtils/CoffFile.cpp:440]: Sounds good. But I would at least add a comment regarding your second part.

**Review by florian-kuebler — APPROVED** [review]

File: src/ObjectUtils/CoffFile.cpp (hunk 2 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp:440]: The way this function is laid out right now, it's also possible to use it "standalone". The explanation I gave you will be in the part that puts this and the unwind info together, i.e., a few pull requests from now.

**Pull request merged**